runtime/src/main/java/org/corfudb/runtime/object/CorfuCompileProxy.java

```diff
@@ -226,6 +228,57 @@ else if (underlyingObject.isOptimisticallyUndoableUnsafe()){
     }
 
 
+    /** Roll the object back to the supplied version if possible.
+     * This function may roll back to a point prior to the requested version.
+     * Otherwise, throws a NoRollbackException.
+     *
+     * Unsafe, requires that the caller has acquired a write lock.
+     *
+     * @param  version              The version to rollback to.
+     * @throws NoRollbackException  If the object cannot be rolled back to
+     *                              the supplied version.
+     */
+    public void rollbackObjectUnsafe(long version) {
+        // If we're already at or before the given version, there's
+        // nothing to do
+        if (underlyingObject.getVersionUnsafe() <= version) {
+            return;
+        }
+
+        final IStreamView sv = underlyingObject.getStreamViewUnsafe();
+        ILogData prevEntry =  sv.current();
+
+        while(prevEntry != null) {
+            if (prevEntry.getType() == DataType.DATA &&
+                    prevEntry.getPayload(rt) instanceof ISMRConsumable) {
+                ISMRConsumable consumable = (ISMRConsumable)
+                        prevEntry.getPayload(rt);
+                List<SMREntry> entries = consumable.getSMRUpdates(streamID);
+                // Make sure all the entries in the set are undoable
+                if (entries.stream().allMatch(x -> x.isUndoable())) {
+                    // start from the end, process one at a time
+                    ListIterator<SMREntry> it = entries.listIterator(entries.size());
+                    while (it.hasPrevious()) {
```

Comment:
wouldn't this fail to undo the first entry in a stream, because it has no 'previous ' ?